File: backend/src/routes/player.js
const express = require('express');
const db = require('../database');
const { authenticateToken } = require('../middleware/auth');

const router = express.Router();

// Get player stats (protected) - auto-creates if not found
router.get('/stats', authenticateToken, (req, res) => {
  try {
    let stats = db.prepare('SELECT * FROM player_stats WHERE user_id = ?').get(req.user.userId);
    
    // Auto-create stats if not found
    if (!stats) {
      const defaultPillars = JSON.stringify({
        craft: { level: 1, xp: 0, xpToNext: 100, name: 'Craft', color: '#06b6d4' },
        vessel: { level: 1, xp: 0, xpToNext: 100, name: 'Vessel', color: '#ec4899' },
        tongue: { level: 1, xp: 0, xpToNext: 100, name: 'Tongue', color: '#8b5cf6' },
        principle: { level: 1, xp: 0, xpToNext: 150, name: 'Principle', color: '#a855f7' },
        capital: { level: 1, xp: 0, xpToNext: 100, name: 'Capital', color: '#10b981' },
      });
      const defaultSkills = JSON.stringify({});
      const defaultActivityStats = JSON.stringify({
        focusSessions: 0,
        focusMinutes: 0,
        habitsCompleted: 0,
        journalEntries: 0,
        questsCompleted: 0,
        meditationMinutes: 0,
      });
      
      const result = db.prepare(
        'INSERT INTO player_stats (user_id, level, xp, xp_to_next, pillars, skills, activity_stats) VALUES (?, 1, 0, 100, ?, ?, ?)'
      ).run(req.user.userId, defaultPillars, defaultSkills, defaultActivityStats);
      
      stats = {
        id: result.lastInsertRowid,
        user_id: req.user.userId,
        level: 1,
        xp: 0,
        xp_to_next: 100,
        pillars: defaultPillars,
        skills: defaultSkills,
        activity_stats: defaultActivityStats,
      };
    }
    
    res.json({
      stats: {
        level: stats.level,
        xp: stats.xp,
        xpToNext: stats.xp_to_next,
        pillars: JSON.parse(stats.pillars),
        skills: JSON.parse(stats.skills),
        activityStats: JSON.parse(stats.activity_stats),
      }
    });
  } catch (error) {
    console.error('Get stats error:', error);
    res.status(500).json({ error: 'Server error' });
  }
});

// Add XP and handle level up (protected)
router.post('/add-xp', authenticateToken, (req, res) => {
  try {
    const { amount, type, skillId } = req.body;
    
    if (!amount || amount <= 0) {
      return res.status(400).json({ error: 'Invalid XP amount' });
    }
    
    // Get current stats
    const stats = db.prepare('SELECT * FROM player_stats WHERE user_id = ?').get(req.user.userId);
    if (!stats) {
      return res.status(404).json({ error: 'Player stats not found' });
    }
    
    let { level, xp, xp_to_next, pillars, skills, activity_stats } = stats;
    pillars = JSON.parse(pillars);
    skills = JSON.parse(skills);
    activity_stats = JSON.parse(activity_stats);
    
    // Add XP
    xp += amount;
    let levelUp = false;
    let levelsGained = 0;
    
    // Check for level up
    while (xp >= xp_to_next) {
      xp -= xp_to_next;
      level += 1;
      levelsGained += 1;
      levelUp = true;
      xp_to_next = Math.floor(xp_to_next * 1.2); // 20% harder each level
    }
    
    // Update pillar XP if type specified
    if (type && pillars[type]) {
      pillars[type].xp += amount;
      if (pillars[type].xp >= pillars[type].xpToNext) {
        pillars[type].xp -= pillars[type].xpToNext;
        pillars[type].level += 1;
        pillars[type].xpToNext = Math.floor(pillars[type].xpToNext * 1.25);
      }
    }
    
    // Update skill XP if skillId specified
    if (skillId && skills[skillId]) {
      skills[skillId].xp += amount;
      if (skills[skillId].xp >= skills[skillId].xpToNext) {
        skills[skillId].xp -= skills[skillId].xpToNext;
        skills[skillId].level += 1;
        skills[skillId].xpToNext = Math.floor(skills[skillId].xpToNext * 1.3);
      }
    }
    
    // Save updated stats
    db.prepare(
      `UPDATE player_stats SET 
        level = ?,
        xp = ?,
        xp_to_next = ?,
        pillars = ?,
        skills = ?,
        activity_stats = ?,
        updated_at = CURRENT_TIMESTAMP
      WHERE user_id = ?`
    ).run(
      level,
      xp,
      xp_to_next,
      JSON.stringify(pillars),
      JSON.stringify(skills),
      JSON.stringify(activity_stats),
      req.user.userId
    );
    
    res.json({
      message: `+${amount} XP earned!`,
      levelUp,
      levelsGained,
      newLevel: level,
      xp,
      xpToNext: xp_to_next,
      progress: Math.floor((xp / xp_to_next) * 100),
    });
  } catch (error) {
    console.error('Add XP error:', error);
    res.status(500).json({ error: 'Server error' });
  }
});

// Update player stats (protected)
router.put('/stats', authenticateToken, (req, res) => {
  try {
    const { level, xp, xpToNext, pillars, skills, activityStats } = req.body;
    
    const result = db.prepare(
      `UPDATE player_stats SET 
        level = COALESCE(?, level),
        xp = COALESCE(?, xp),
        xp_to_next = COALESCE(?, xp_to_next),
        pillars = COALESCE(?, pillars),
        skills = COALESCE(?, skills),
        activity_stats = COALESCE(?, activity_stats),
        updated_at = CURRENT_TIMESTAMP
      WHERE user_id = ?`
    ).run(
      level,
      xp,
      xpToNext,
      pillars ? JSON.stringify(pillars) : null,
      skills ? JSON.stringify(skills) : null,
      activityStats ? JSON.stringify(activityStats) : null,
      req.user.userId
    );
    
    if (result.changes === 0) {
      return res.status(404).json({ error: 'Player stats not found' });
    }
    
    res.json({ message: 'Stats updated successfully' });
  } catch (error) {
    console.error('Update stats error:', error);
    res.status(500).json({ error: 'Server error' });
  }
});

// Track activity (protected)
router.post('/activity', authenticateToken, (req, res) => {
  try {
    const { type, count = 1 } = req.body;
    
    const validTypes = ['focusSessions', 'focusMinutes', 'habitsCompleted', 'journalEntries', 'questsCompleted', 'meditationMinutes'];
    if (!validTypes.includes(type)) {
      return res.status(400).json({ error: 'Invalid activity type' });
    }
    
    // Get current stats
    const stats = db.prepare('SELECT activity_stats FROM player_stats WHERE user_id = ?').get(req.user.userId);
    if (!stats) {
      return res.status(404).json({ error: 'Player stats not found' });
    }
    
    const activityStats = JSON.parse(stats.activity_stats);
    activityStats[type] = (activityStats[type] || 0) + count;
    
    // Save updated activity stats
    db.prepare(
      'UPDATE player_stats SET activity_stats = ?, updated_at = CURRENT_TIMESTAMP WHERE user_id = ?'
    ).run(JSON.stringify(activityStats), req.user.userId);
    
    res.json({
      message: 'Activity tracked',
      type,
      total: activityStats[type],
    });
  } catch (error) {
    console.error('Track activity error:', error);
    res.status(500).json({ error: 'Server error' });
  }
});

module.exports = router;

File: services/gateway/src/handlers/player.rs
//! Player stats handlers with PostgreSQL integration

use actix_web::{HttpRequest, HttpResponse, web};
use serde::Deserialize;
use serde_json::json;
use sqlx::PgPool;
use uuid::Uuid;

use crate::handlers::auth::get_request_id;
use crate::models::{ActivityStats, AddXPResponse, PlayerStats};
use crate::services::AuthService;

/// Extract user ID from JWT token in request
fn extract_user_id(req: &HttpRequest, auth_service: &AuthService) -> Option<Uuid> {
    let token = req
        .headers()
        .get("authorization")
        .and_then(|h| h.to_str().ok())
        .and_then(|h| h.strip_prefix("Bearer "));

    token.and_then(|t| {
        auth_service
            .validate_token(t)
            .ok()
            .and_then(|claims| Uuid::parse_str(&claims.sub).ok())
    })
}

/// Get player stats from PostgreSQL
pub async fn get_stats(
    req: HttpRequest,
    pool: web::Data<PgPool>,
    auth_service: web::Data<AuthService>,
) -> HttpResponse {
    let request_id = get_request_id(&req);

    let user_id = match extract_user_id(&req, &auth_service) {
        Some(id) => id,
        None => {
            return HttpResponse::Unauthorized().json(json!({
                "status": "error",
                "error": { "code": "UNAUTHORIZED", "message": "Authentication required" }
            }));
        }
    };

    let result = sqlx::query_as::<_, (i64, i32, i32, serde_json::Value, serde_json::Value)>(
        r#"
        SELECT gs.xp_total as xp, gs.level, gs.streak_days,
               COALESCE((
                   SELECT json_object_agg(domain, json_build_object(
                       'level', 1,
                       'xp', 0,
                       'xpToNext', 100,
                       'name', domain,
                       'color', '#00ff88'
                   ))
                   FROM (
                       SELECT 'body' as domain
                       UNION SELECT 'mind'
                       UNION SELECT 'german'
                       UNION SELECT 'code'
                       UNION SELECT 'finance'
                   ) domains
               ), '{}'::json) as pillars,
               '{}'::json as skills
        FROM gamification_stats gs
        WHERE gs.user_id = $1
        "#,
    )
    .bind(user_id)
    .fetch_one(pool.get_ref())
    .await;

    match result {
        Ok((xp, level, _streak, pillars, skills)) => {
            let xp_to_next = level * 100;

            let stats = PlayerStats {
                level,
                xp: xp as i32,
                xp_to_next,
                pillars,
                skills,
                activity_stats: ActivityStats::default(),
            };

            HttpResponse::Ok().json(json!({
                "status": "success",
                "data": { "stats": stats },
                "meta": { "request_id": request_id }
            }))
        }
        Err(e) => {
            tracing::warn!("Failed to get player stats: {}", e);
            HttpResponse::Ok().json(json!({
                "status": "success",
                "data": {
                    "stats": PlayerStats {
                        level: 1,
                        xp: 0,
                        xp_to_next: 100,
                        pillars: json!({}),
                        skills: json!({}),
                        activity_stats: ActivityStats::default(),
                    }
                },
                "meta": { "request_id": request_id }
            }))
        }
    }
}

/// Request body for adding XP
#[derive(Debug, Deserialize)]
pub struct AddXPRequest {
    pub amount: i32,
    pub pillar: Option<String>,
    pub skill_id: Option<String>,
}

/// Add XP to player
pub async fn add_xp(
    req: HttpRequest,
    body: web::Json<AddXPRequest>,
    pool: web::Data<PgPool>,
    auth_service: web::Data<AuthService>,
) -> HttpResponse {
    let request_id = get_request_id(&req);

    let user_id = match extract_user_id(&req, &auth_service) {
        Some(id) => id,
        None => {
            return HttpResponse::Unauthorized().json(json!({
                "status": "error",
                "error": { "code": "UNAUTHORIZED", "message": "Authentication required" }
            }));
        }
    };

    let AddXPRequest {
        amount,
        pillar: _pillar,
        skill_id: _skill_id,
    } = body.into_inner();

    // Get current stats for user
    let current = sqlx::query_as::<_, (i64, i32)>(
        "SELECT xp_total, level FROM gamification_stats WHERE user_id = $1",
    )
    .bind(user_id)
    .fetch_one(pool.get_ref())
    .await;

    let (current_xp, current_level) = match current {
        Ok((xp, lvl)) => (xp as i32, lvl),
        Err(_) => (0, 1),
    };

    let new_xp = current_xp + amount;
    let new_level = (new_xp / 100) + 1;
    let xp_to_next = new_level * 100;
    let progress = (new_xp % 100) as f64 / 100.0;
    let level_up = new_level > current_level;
    let levels_gained = new_level - current_level;

    // Update in database
    let _ = sqlx::query(
        "UPDATE gamification_stats SET xp_total = $1, level = $2, updated_at = NOW() WHERE user_id = $3"
    )
    .bind(new_xp as i64)
    .bind(new_level)
    .bind(user_id)
    .execute(pool.get_ref())
    .await;

    HttpResponse::Ok().json(json!({
        "status": "success",
        "data": AddXPResponse {
            message: format!("Added {} XP", amount),
            level_up,
            levels_gained,
            new_level,
            xp: new_xp,
            xp_to_next,
            progress,
        },
        "meta": { "request_id": request_id }
    }))
}

/// Track activity
#[derive(Debug, Deserialize)]
pub struct TrackActivityRequest {
    pub activity_type: String,
    pub count: Option<i32>,
}

pub async fn track_activity(
    req: HttpRequest,
    body: web::Json<TrackActivityRequest>,
) -> HttpResponse {
    let request_id = get_request_id(&req);

    HttpResponse::Ok().json(json!({
        "status": "success",
        "data": {
            "message": "Activity tracked",
            "type": body.activity_type,
            "total": body.count.unwrap_or(1),
        },
        "meta": { "request_id": request_id }
    }))
}

/// Update player stats
#[derive(Debug, Deserialize)]
pub struct UpdateStatsRequest {
    pub level: Option<i32>,
    pub xp: Option<i32>,
}

pub async fn update_stats(
    req: HttpRequest,
    body: web::Json<UpdateStatsRequest>,
    pool: web::Data<PgPool>,
    auth_service: web::Data<AuthService>,
) -> HttpResponse {
    let request_id = get_request_id(&req);

    let user_id = match extract_user_id(&req, &auth_service) {
        Some(id) => id,
        None => {
            return HttpResponse::Unauthorized().json(json!({
                "status": "error",
                "error": { "code": "UNAUTHORIZED", "message": "Authentication required" }
            }));
        }
    };

    if let Some(xp) = body.xp {
        let _ = sqlx::query("UPDATE gamification_stats SET xp_total = $1 WHERE user_id = $2")
            .bind(xp as i64)
            .bind(user_id)
            .execute(pool.get_ref())
            .await;
    }

    if let Some(level) = body.level {
        let _ = sqlx::query("UPDATE gamification_stats SET level = $1 WHERE user_id = $2")
            .bind(level)
            .bind(user_id)
            .execute(pool.get_ref())
            .await;
    }

    HttpResponse::Ok().json(json!({
        "status": "success",
        "data": { "message": "Stats updated" },
        "meta": { "request_id": request_id }
    }))
}
